// fs/lib/file.js
/**
 * Usage: 用于IDE的智能提示
 * Author: Spikef < Spikef@Foxmail.com >
 * Copyright: Envirs Team < http://envirs.com >
 */

var obj = new ActiveXObject('Scripting.FileSystemObject');

var file = function(path) {
    var f = obj.GetFile(path);

    return {
        // Properties
        Attributes: f.Attributes,                   // 设置或返回指定文件的属性。
        DateCreated: f.DateCreated,                 // 返回指定文件创建的日期和时间。
        DateLastAccessed: f.DateLastAccessed,	    // 返回指定文件最后被访问的日期和时间。
        DateLastModified: f.DateLastModified,	    // 返回指定文件最后被修改的日期和时间。
        Drive: f.Drive,         	                // 返回指定文件或文件夹所在的驱动器的驱动器字母。
        Name: f.Name,                               // 设置或返回指定文件的名称。
        ParentFolder: f.ParentFolder,	            // 返回指定文件或文件夹的父文件夹对象。
        Path: f.Path,	                            // 返回指定文件的路径。
        ShortName: f.ShortName,	                    // 返回指定文件的短名称（8.3 命名约定）。
        ShortPath: f.ShortPath,	                    // 返回指定文件的短路径（8.3 命名约定）。
        Size: f.Size,	                            // 返回指定文件的尺寸（字节）。
        Type: f.Type,	                            // 返回指定文件的类型。

        // Methods
        SetAttr: function(value) {
            f.Attributes = value;
        },
        Rename: function(value) {
            f.Name = value;
        },

        /**
         * 把指定文件从一个位置拷贝到另一个位置。
         * @param destination: 必需的。复制文件或文件夹的目的地。不允许使用通配符。
         * @param overwrite: 可选的。指示是否可覆盖已有的文件或文件夹的布尔值。True 表示文件/文件夹可被覆盖，false 表示文件/文件夹不可以被覆盖。默认是 true。
         */
        Copy: function(destination, overwrite) {
            overwrite = overwrite === false ? false : true;

            f.Copy(destination, overwrite);
        },

        /**
         * 把指定文件从一个位置移动到另一个位置。
         * @param destination: 必需的。移动文件或文件夹的目的地。
         */
        Move: function(destination) {
            f.Move(destination);
        },

        /**
         * 删除指定文件。
         * @param force: 可选的。指示是否可删除只读文件或文件夹的布尔值。True 指示只读文件/文件夹可被删除，false 指示文件/文件夹不可被删除。默认是 false。
         * @throws 若指定文件或文件夹不存在，错误发生。
         */
        Delete: function(force) {
            force = force === true ? true : false;

            f.Delete(force);
        }
    }
};

module.exports = file;
